商品出品機能の実装

akirausui19871117 opened the pull request:
お世話になっております！商品出品機能の実装を行いましたので、コードレビューをお願いいたします。

What
商品出品機能の実装
why
ログインしたユーザーが商品を問題なく出品できるようにするため。

akirausui19871117 committed:
ルーティングの設定〜アクションの設定

出品ページへの遷移先の変更
ルーティングの設定
アクションの設定
itemsモデルの作成とバリデーションの記述
アソシエーションの記述

akirausui19871117 committed:
商品出品機能の実装

akirausui19871117 committed:
テストコードの記述

akirausui19871117 commented:
ログイン状態のユーザーは、商品出品ページへ遷移できる動画 
https://gyazo.com/7baf5134bfc2e8878ced3f07422f936c

金額が入力されると同時に、販売手数料と販売利益が表示される動画
https://gyazo.com/6f589a751d4319484fcc0b93418e3c59

正しく情報を記入すると、商品を出品できる動画
https://gyazo.com/08f3d8ad53590cee22ace4aae5734857

入力に問題のある状態では商品は出品できず、エラーメッセージが出力される動画
https://gyazo.com/4932ef27052658cfbd41f5848a11eae9

ログアウト状態のユーザーは、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
https://gyazo.com/1f419581c57541ef3d253b0ba0dda517

テスト結果の画像
https://gyazo.com/0502cb3405946d85936c8afd50062d01

maimai0401 commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,27 @@
 class ItemsController < ApplicationController
+  before_action :authenticate_user!, only: [:new,:edit]
```
```
  before_action :authenticate_user!, only: [:new,:edit]
```
editアクションに関しましては、今後の実装で記述を行うこととなりますので現段階では不要です。
削除かコメントアウトをしておきましょう。

また、より保守性の高いアプリとするためcreateアクションを追加しましょう。

maimai0401 commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,27 @@
+
+private
+  def item_params
+     params.require(:item).permit(:image,:name,:explanation,:category_id,:condition_id,:delivery_fee_id,:prefecture_id,:delivery_day_id,:price)
```
現在の記述内容ですと、出品者の情報が登録されていない状態ではないでしょうか？
出品者の情報も登録されるように記述を見直してみましょう。

maimai0401 commented on app/javascript/item_price.js:
```diff
@@ -0,0 +1,14 @@
+window.addEventListener('load',() => {
+  console.log("OK");
```
```
  console.log("OK");
  console.log(inputValue);
  console.log(addTaxDom);
  console.log(addProfit);
```
デバッグ用のコードは削除しましょう。

maimai0401 commented on app/models/category.rb:
```diff
@@ -0,0 +1,18 @@
+      { id: 11, name: 'その他'}
+    ]
+  include ActiveHash::Associations
+  belongs_to :item
```
今回の実装では、商品とカテゴリーの関係は1対多の関係となりそうです。
アソシエーション を見直しましょう。

また、その他のactive_hashを用いたファイルのアソシエーション の記述も見直しましょう。

maimai0401 commented on app/models/item.rb:
```diff
@@ -0,0 +1,19 @@
+  belongs_to :user, optional: true
+  has_one :buy
+  has_one_attached :image
+
```
予期せぬエラーを防ぐため、active_hashとのアソシエーション を追記しましょう。

maimai0401 commented on app/views/items/new.html.erb:
```diff
@@ -5,10 +5,10 @@
-    <%= form_with local: true do |f| %>
+    <%= form_with model:@item, local: true do |f| %>
 
     <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
上記のようなコメントアウトは実装方針の内容であり、企業にポートフォリオとして提出する際に、好ましくないため削除しましょう。

maimai0401 commented on spec/factories/items.rb:
```diff
@@ -0,0 +1,17 @@
+  factory :item do
+    name                   { Faker::Name.initials(number: 6) }
+    explanation            { Faker::Lorem.sentence}
+    price                  { "10000" }
```
```
   price                  { "10000" }
    category_id            { "2" }
    condition_id           { "2" }
    delivery_fee_id        { "2" }
    prefecture_id          { "2" }
    delivery_day_id        { "2" } 
```
integer型に関するテストとなりますので、””は外しておきましょう。

maimai0401 commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,66 @@
+        expect(@item.errors.full_messages).to include("Explanation can't be blank")
+      end
+
+      it 'category_idが--では投稿できない' do
```
```
 it 'category_idが--では投稿できない' do
        @item.category_id = '1'
        @item.valid?
        expect(@item.errors.full_messages).to include("Category Select")

      end

      it 'condition_idが--では投稿できない' do
        @item.condition_id = '1'
        @item.valid?
        expect(@item.errors.full_messages).to include("Condition Select")

      end

      it 'delivery_fee_idが--では投稿できない' do
        @item.delivery_fee_id = '1'
        @item.valid?
        expect(@item.errors.full_messages).to include("Delivery fee Select")
      end

      it 'prefecture_idが--では投稿できない' do
        @item.prefecture_id = '1'
        @item.valid?
        expect(@item.errors.full_messages).to include("Prefecture Select")
      end

      it 'delivery_day_idが--では投稿できない' do
        @item.delivery_day_id = '1'
        @item.valid?
        expect(@item.errors.full_messages).to include("Delivery day Select")
      end
```
integer型に関するテストとなりますので、””は外しておきましょう。

maimai0401 commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,66 @@
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Delivery day Select")
+      end
+
```
より正確なテストを行うために、下記テストコードを追加で実装しましょう。
・価格は空では登録できないこと
・価格は全角文字では登録できないこと
・価格は半角英数混合では登録できないこと
・価格は半角英語だけでは登録できないこと
・価格は299円以下では登録できないこと
・価格は10,000,000以上では登録できないこと

maimai0401 reviewed (commented):
レビュー依頼ありがとうございます。
９点コメントいたしましたので、ご確認をお願いいたします。
修正後、再度レビュー依頼フォームにてご提出くださいませ。

akirausui19871117 committed:
修正その１

Active Hashで記載した項目とitemモデルのアソシエーションの記述。
不要な記述の削除
FactoryBotのinterger型の項目の数値化
テストコードの追記

akirausui19871117 commented:
ご指摘いただきましてありがとうございました。

修正を加えましたので、再度コードレビューをしていただきますようお願いいたします。

NaturalRadio commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,27 @@
+  end
+
+  def create
+    @item = Item.create(item_params)
+     if @item.save
```
ActiveRecordのcreateメソッドは、newメソッドとsaveメソッドを同時に行うため、
現在の記述では、13行目と14行目で一部重複することになります。
そのため、今回は13行目でcreateメソッドではなくnewメソッドを使用してあげると良さそうです。

NaturalRadio commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,95 @@
+        expect(@item.errors.full_messages).to include("Price Out of setting range")
+      end
+
+      it '価格は半角英語だけでは登録できないこと' do
```
```
価格は半角英数混合では登録できないこと
```
上記テスト項目を追加しておきましょう！

NaturalRadio reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします！

akirausui19871117 commented:
ご指摘いただきましてありがとうございました！
いただきましたレビューに従い、修正を加えましたので、御確認頂ますようお願いいたします！

akirausui19871117 committed:
修正その２

テストコードの追加
アソシエーションの記述の変更
アイテムコントローラーの記述の修正（createアクションのインスタンスを.createから.newへ変更した）

NaturalRadio commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,27 @@
+  before_action :authenticate_user!, only: [:new,:edit,:create]
+
   def index
+    @items = Item.all
```
```
@items = Item.all
```
こちら一覧機能にて確認をさせていただきたい記述となりますのでコメントアウトをお願いします。

NaturalRadio commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,102 @@
+
+
+      it '価格は299円以下では登録できないこと' do
+        @item.price = '299'
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Price Out of setting range")
+      end
+
+      it '価格は10,000,000以上では登録できないこと' do
+        @item.price = '10000000'
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Price Out of setting range")
+      end
```
こちらは数値を代入させたいため、`'(シングルクォーテーション)`は削除しておきましょう！

NaturalRadio reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
先程指摘漏れがございましたので追加でコメントしております。
お手数をお掛けしますが、ご確認お願いします！

akirausui19871117 commented:
再度ご指摘いただきましてありがとうございました。
追記に従い、修正を加えましたので、再度コードレビューしていただけると幸いです！
よろしくお願いいたします！

akirausui19871117 committed:
修正その３

テストコードの数値部分のシングルクォーテーションを削除
Item.allのコメントアウト

kuu0123 commented on config/routes.rb:
```diff
@@ -1,5 +1,6 @@
 Rails.application.routes.draw do
   root to: "items#index"
   devise_for :users
+  resources :items, only: [:index, :new, :create,:edit]
```
```ruby
resources :items, only: [:index, :new, :create,:edit]
```
今回は商品出品機能の実装のプルリクエストですので、上記の記述からeditの記述は削除しておきましょう。
実装に必要なルーティングのみ生成するという観点からも修正が必要ですが、
上記を記載したままの場合、商品編集機能実装時にroutes.rbの差分が出ず適切なコードレビューが受けられない可能性がございます。

kuu0123 reviewed (commented):
レビュー依頼ありがとうございます！

修正点確認させていただきました。
細かい箇所にはなりますが、追加で1点コメントいたしましたので、ご確認お願いします！

akirausui19871117 commented:
ご指摘いただきましてありがとうございました！
再度修正いたしましたので御確認頂ますようお願いいたします！

akirausui19871117 committed:
修正その４

ルーティングの修正（不要な記述の削除）

kuu0123 commented:
レビュー依頼ありがとうございます！

商品出品機能、LGTMとさせていただきます！実装お疲れさまでした！

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に次の実装のためのブランチを切るようにしましょう。
引き続き実装頑張ってください！
